### PR TITLE
Decouple Metrics Service from MobileCore and Replace Callbacks by Rx

### PR DESCRIPTION
**Pull request by danielpassos**
## Related issues

* [AGDROID-796](https://issues.jboss.org/browse/AGDROID-796)
* [AGDROID-814](https://issues.jboss.org/browse/AGDROID-814)

### Decouple Metrics Service from MobileCore

Replace:

```java
MetricsService metricsService = MobileCore.getInstance().getService(MetricsService.class);
```

By

```java
MetricsService metricsService = MobileCore.getInstance().getMetricsService();
```

### Replace Callback by Rx

Replace:

```java
metricsService.publish(theType, theMetrics, theCallback));
```

By

```java
metricsService.publish(theType, theMetrics)
    .requestOn(new AppExecutors().mainThread())
    .respondWith(new Responder<Boolean>() {
        @Override
        public void onResult(Boolean value) {
            // Yay
        }

        @Override
        public void onException(Exception exception) {
            // Oops
        }
    });
```

### CODE REVIEW

File: core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java [hunk 1]
```diff
@@ -74,25 +43,60 @@ public void sendAppAndDeviceMetrics(final Callback callback) {
      * @param type type of the enclosing metrics event
      * @param metrics Metrics to send
      */
-    public void publish(String type, Metrics... metrics) {
-        publish(type, metrics, null);
+    public Request<Boolean> publish(@NonNull String type, @NonNull final Metrics... metrics) {
+
+        return Requester.call(() -> MobileCore.getInstance().getHttpLayer().newRequest())
```
Comment (secondsun) [inline core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java]: You should be able to use it like this : 

```
return MobileCore.getInstance().getHttpLayer().newRequest()
                                        .post(url, createMetricsJSONObject(type, metrics).toString().getBytes())
                                        .requestMap(httpResponse -> 
                                            switch (httpResponse.getStatus()) {
                                                case HTTP_OK:
                                                    return Boolean.TRUE;
                                                default:
                                                    throw (new HttpException(
                                                                    httpResponse.getStatus()));
                                            }
                                        ).requestOn(new AppExecutors().networkThread()));
```

I'll need to pop this into android studio and confirm the exact syntac, but you don't need to wrap everything in a request when you can just chain them.

File: core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java [hunk 2]
```diff
@@ -1,71 +1,40 @@
...
-        return true;
+    public MetricsService(String url) {
+        this.url = url;
     }
 
     /**
      * Send default metrics
      */
-    public void sendAppAndDeviceMetrics() {
+    public Request<Boolean> sendAppAndDeviceMetrics() {
```
Comment (secondsun) [inline core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java]: missing @return in javadoc

File: core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java [hunk 3]
```diff
@@ -74,25 +43,60 @@ public void sendAppAndDeviceMetrics(final Callback callback) {
      * @param type type of the enclosing metrics event
      * @param metrics Metrics to send
      */
```
Comment (secondsun) [inline core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java]: missing @return in javadoc

File: core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java (hunk 1 above)
Comment (danielpassos) [reply core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java]: For some reason `.requestMap(httpResponse` from `.post` does not work for me.

Comment (secondsun) [reply core/src/main/java/org/aerogear/mobile/core/metrics/MetricsService.java]: Ahhh yeah looks like RequestMapFunction has silly return type. that I need to fix

**Comment by secondsun** [comment]
Works on my machine

**Comment by ciaranRoche** [comment]
@wtrocki awesome thanks, between the decoupling of push and metrics, and the whitelisting/blacklisting of browsers there is some really cool changes in the next release.

**Comment by wtrocki** [comment]
Merging to perform tests with @darahayes

**Comment by danielpassos** [comment]
@ciaranRoche You are free to do a release whenever you want. For now, we (@secondsun and I) are doing it from time to time on Fridays using 0.1.0-[year]-[week-of-the-year]

Here is a [doc](https://github.com/aerogear/aerogear-android-sdk/blob/master/docs/contrib/releases.adoc) @darahayes wrote to help you on the release process.

**Pull request merged**